File: backend/app/models/fitness.py
"""Pydantic Schemas für Fitness-Score API.

Basiert auf dem Banister Fitness-Fatigue-Modell:
- TRIMP (Edwards): Trainingsbelastung pro Session
- CTL (Chronic Training Load): Fitness (~42 Tage EWMA)
- ATL (Acute Training Load): Ermüdung (~7 Tage EWMA)
- TSB (Training Stress Balance): Form = CTL - ATL
- ACWR (Acute:Chronic Workload Ratio): Verletzungsrisiko
"""

from __future__ import annotations

from pydantic import BaseModel, Field


class FormIndicatorResponse(BaseModel):
    """Frische/Ermüdungs-Indikator basierend auf TSB."""

    status: str = Field(..., description="fresh | normal | fatigued")
    label: str = Field(..., description="Frisch | Normal | Ermüdet")
    color: str = Field(..., description="green | yellow | orange")
    recommendation: str


class ACWRResponse(BaseModel):
    """Acute:Chronic Workload Ratio — Verletzungsrisiko-Indikator."""

    ratio: float
    zone: str = Field(..., description="low | optimal | warning | danger")
    message: str


class FitnessScoreResponse(BaseModel):
    """Haupt-Response: Aktueller Fitness-Score mit allen Indikatoren."""

    score: int = Field(..., ge=0, le=100)
    endurance_score: int = Field(..., ge=0, le=100)
    strength_score: int = Field(..., ge=0, le=100)
    trend: str = Field(..., description="rising | stable | falling")
    trend_label: str = Field(..., description="↑ steigend | → stabil | ↓ fallend")
    form: FormIndicatorResponse
    acwr: ACWRResponse | None = None
    context_message: str
    updated_at: str


class FitnessDataPoint(BaseModel):
    """Ein Datenpunkt im Fitness-Verlauf (für Charts)."""

    date: str
    value: float


class FitnessHistoryResponse(BaseModel):
    """Fitness-Verlauf über Zeit (für Fortschritt-Charts)."""

    ctl_history: list[FitnessDataPoint]
    atl_history: list[FitnessDataPoint]
    tsb_history: list[FitnessDataPoint]
    score_history: list[FitnessDataPoint]


class RecalculateResponse(BaseModel):
    """Response für Batch-Neuberechnung."""

    recalculated_sessions: int


# ---------------------------------------------------------------------------
# Insight-Engine Responses
# ---------------------------------------------------------------------------


class InsightResponse(BaseModel):
    """Ein einzelner proaktiver Hinweis."""

    type: str = Field(..., description="warning | trend | achievement | recommendation | info")
    priority: int = Field(..., ge=1, le=10)
    title: str
    message: str
    category: str = Field(..., description="load | balance | performance | plan | recovery")
    icon: str


class InsightsListResponse(BaseModel):
    """Liste aktiver Insights."""

    insights: list[InsightResponse]
    generated_at: str


class IntensityDistributionResponse(BaseModel):
    """80/20-Verteilung der Trainingsintensität."""

    low_percent: float
    medium_percent: float
    high_percent: float
    is_polarized: bool
    total_minutes: float


class TrainingQualityResponse(BaseModel):
    """Trainingsqualität-Metriken."""

    intensity_distribution: IntensityDistributionResponse
    monotony: float
    monotony_level: str
    strain: float
    strain_level: str


# ---------------------------------------------------------------------------
# Today-Dashboard Responses
# ---------------------------------------------------------------------------


class LastSessionSummary(BaseModel):
    """Zusammenfassung der letzten Session mit Einordnung."""

    id: int
    date: str
    workout_type: str
    training_type: str | None = None
    distance_km: float | None = None
    duration_seconds: int | None = None
    avg_pace_formatted: str | None = None
    avg_heartrate: float | None = None
    exercise_count: int | None = None
    tonnage_kg: float | None = None
    rpe: float | None = None
    trimp_score: float | None = None
    comparison_message: str = ""


class DayStatus(BaseModel):
    """Status eines einzelnen Tages in der Wochenübersicht."""

    date: str
    day_name: str
    has_planned: bool
    has_completed: bool
    status: str = Field(..., description="completed | planned | skipped | extra | rest")


class WeekProgressResponse(BaseModel):
    """Wochenfortschritt: geplant vs. erledigt."""

    sessions_completed: int
    sessions_planned: int
    distance_completed_km: float
    distance_planned_km: float | None = None
    time_completed_seconds: int
    time_planned_seconds: int | None = None
    days: list[DayStatus]


class NextSessionInfo(BaseModel):
    """Nächste geplante Session aus dem Wochenplan."""

    day_name: str = Field(..., description="z.B. 'Morgen', 'Mittwoch'")
    workout_type: str = Field(..., description="running | strength")
    description: str = Field(..., description="z.B. 'Intervall-Lauf' oder 'Oberkörper'")


class GoalSummary(BaseModel):
    """Zusammenfassung des aktiven Trainingsziels."""

    title: str
    days_until: int
    target_time_formatted: str | None = None


class TodayResponse(BaseModel):
    """Aggregierte Dashboard-Daten für die Heute-Seite."""

    greeting: str
    motivation: str | None = None
    fitness_score: FitnessScoreResponse
    last_session: LastSessionSummary | None = None
    week_progress: WeekProgressResponse
    insights: list[InsightResponse]
    next_session: NextSessionInfo | None = None
    goal_summary: GoalSummary | None = None

File: frontend/src/api/fitness.ts
import { useQuery } from '@tanstack/react-query';
import { apiClient } from './client';

// ---------------------------------------------------------------------------
// Types — spiegeln die Backend-Pydantic-Schemas wider
// ---------------------------------------------------------------------------

export interface FormIndicatorResponse {
  status: 'fresh' | 'normal' | 'fatigued';
  label: string;
  color: 'green' | 'yellow' | 'orange';
  recommendation: string;
}

export interface ACWRResponse {
  ratio: number;
  zone: 'low' | 'optimal' | 'warning' | 'danger';
  message: string;
}

export interface FitnessScoreResponse {
  score: number;
  endurance_score: number;
  strength_score: number;
  trend: 'rising' | 'stable' | 'falling';
  trend_label: string;
  form: FormIndicatorResponse;
  acwr: ACWRResponse | null;
  context_message: string;
  updated_at: string;
}

export interface FitnessDataPoint {
  date: string;
  value: number;
}

export interface FitnessHistoryResponse {
  ctl_history: FitnessDataPoint[];
  atl_history: FitnessDataPoint[];
  tsb_history: FitnessDataPoint[];
  score_history: FitnessDataPoint[];
}

export interface InsightResponse {
  type: 'warning' | 'trend' | 'achievement' | 'recommendation' | 'info';
  priority: number;
  title: string;
  message: string;
  category: 'load' | 'balance' | 'performance' | 'plan' | 'recovery';
  icon: string;
}

export interface LastSessionSummary {
  id: number;
  date: string;
  workout_type: string;
  training_type: string | null;
  distance_km: number | null;
  duration_seconds: number | null;
  avg_pace_formatted: string | null;
  avg_heartrate: number | null;
  exercise_count: number | null;
  tonnage_kg: number | null;
  rpe: number | null;
  trimp_score: number | null;
  comparison_message: string;
}

export interface DayStatus {
  date: string;
  day_name: string;
  has_planned: boolean;
  has_completed: boolean;
  status: 'completed' | 'planned' | 'skipped' | 'extra' | 'rest';
}

export interface WeekProgressResponse {
  sessions_completed: number;
  sessions_planned: number;
  distance_completed_km: number;
  distance_planned_km: number | null;
  time_completed_seconds: number;
  time_planned_seconds: number | null;
  days: DayStatus[];
}

export interface NextSessionInfo {
  day_name: string;
  workout_type: 'running' | 'strength';
  description: string;
}

export interface GoalSummary {
  title: string;
  days_until: number;
  target_time_formatted: string | null;
}

export interface TodayResponse {
  greeting: string;
  motivation: string | null;
  fitness_score: FitnessScoreResponse;
  last_session: LastSessionSummary | null;
  week_progress: WeekProgressResponse;
  insights: InsightResponse[];
  next_session: NextSessionInfo | null;
  goal_summary: GoalSummary | null;
}

// ---------------------------------------------------------------------------
// API-Funktionen
// ---------------------------------------------------------------------------

export async function getToday(): Promise<TodayResponse> {
  const response = await apiClient.get<TodayResponse>('/api/v1/fitness/today');
  return response.data;
}

export async function getFitnessScore(): Promise<FitnessScoreResponse> {
  const response = await apiClient.get<FitnessScoreResponse>('/api/v1/fitness/score');
  return response.data;
}

export async function getFitnessHistory(days: number = 90): Promise<FitnessHistoryResponse> {
  const response = await apiClient.get<FitnessHistoryResponse>(
    `/api/v1/fitness/history?days=${days}`,
  );
  return response.data;
}

// ---------------------------------------------------------------------------
// React Query Hooks
// ---------------------------------------------------------------------------

export function useToday() {
  return useQuery({
    queryKey: ['fitness', 'today'],
    queryFn: getToday,
    staleTime: 3 * 60 * 1000, // 3 Minuten — Dashboard täglich aktuell
  });
}

export function useFitnessHistory(days: number = 90) {
  return useQuery({
    queryKey: ['fitness', 'history', days],
    queryFn: () => getFitnessHistory(days),
    staleTime: 10 * 60 * 1000,
  });
}
